### ISSUE TITLE
twisted.spread.pb.IUsernameMD5Password's docstring should say it accepts an MD5 digest of a password, not plaintext

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3056]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3056|
|Type|defect|
|Created|2008-02-27 14:11:20Z|
|Branch|https://github.com/twisted/twisted/tree/pb-md5-docstring-3056|

This interface seems to be entirely redundant with `IUsernameHashedPassword`.

Attachments:

* [IUsernameMD5Password_Deprecated.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b0b/b0be911eadc924612a478b7fe67e2d9f1b7d0bc9/9f63f30bf360712dce66a8c62395e55cf16ffabe.patch) (1903 bytes) - added by jesstess on 2009-12-21 19:54:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3056 3056
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_pb_md5_docstring_3056 branches/pb-md5-docstring-3056
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1204121480000000 1204121480000000
changetime__1266598333000000 1266598333000000
version__None None
owner__ 
cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess removed owner|
|-|-|

Some questions on this ticket:
 * I'm not sure how to eke a DeprecationWarning out of an interface to make a test case. Implementing the interface doesn't cause one. Is there a standard way to do this?
 * Should I be doing something to class _PortalAuthChallenger as well since it implements IUsernameMD5Password?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @jesstess|
|-|-|

  1. `twisted.python.deprecate.deprecatedModuleAttributes` will let you deprecate the attribute (this results in a deprecation warning when the attribute is accessed, either via '.' or an import statement).  I'm not sure if there's a way to hook into `implements()` or the other zope.interface declaration APIs; that'd be cool, but probably isn't totally necessary.  Emitting a deprecation from `__init__` won't do anything, I think, as interfaces aren't intended to be instantiated (and perhaps **cannot** be instantiated (I hope)).
  1. Yea.  It'd be good to eliminate any uses of this interface within Twisted, as long as doing so doesn't introduce backwards compatibility issues.  In the case of `_PortalAuthChallenger`, this is a private class, so we can do pretty much anything we want to it without worrying about compatibility issues.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess commented|
|-|-|

(In [28129]) Branching to 'pb-md5-deprecate-3056'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess commented|
|-|-|

(In [28132]) Deprecate IUsernameMD5Password using deprecatedModuleAttribute and add
a unit test for the deprecation.

refs [#3056](https://github.com/twisted/twisted/issues/3056)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

(In [28133]) Remove references to IUsernameMD5Password in _PortalAuthChallenger.

refs [#3056](https://github.com/twisted/twisted/issues/3056)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess removed owner|
|-|-|

I left checkPassword and checkMD5Password as distinct functions in _PortalAuthChallenger to be conservative about compatibility, but I can roll checkMD5Password into checkPassword if people prefer that.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

Can you open a ticket for the removal of this interface?

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess removed owner|
|-|-|

[#4267](https://github.com/twisted/twisted/issues/4267) was opened for the removal.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @jesstess|
|-|-|

  1. There's some weird quoting in `test_pb.py`, in the docstring for the new test, `test_IUsernameMD5Password`.
  1. You can just do `pb.IUsernameMD5Password` in that test, rather than using getattr, can't you?  It should do the same thing, I think, but it's simpler.
  1. I suppose that `_PortalAuthChallenger.checkMD5Password` should be deprecated as well.  Someone might get their hands on an instance of `_PortalAuthChallenger` via cred (eg, by implementing a checker for it) and call the method.  They should get warned in that case.  As long as we're doing that, I guess there's also no reason to remove the `IUsernameMD5Password` implements declaration from `_PortalAuthChallenger` right now either.  We can delete it when we delete the interface definition (and delete `checkMD5Password` then as well).  I think the only annoying thing about this is that to avoid a deprecation warning when calling `checkPassword` the implementation will have to shift around a bit.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

Errgh.  In responding to a recent post on tpml about this code, I realized that `IUsernameHashedPassword` and `IUsernameMD5Password` actually have a difference.  `IUsernameMD5Password.checkMD5Password` accepts an MD5 digest of a password, while `IUsernameHashedPassword.checkPassword` accepts a plaintext password.  Thus, if you don't want to store plaintext passwords on your server, `IUsernameMD5Password` is a better choice.

I think was misled by the documentation for `IUsernameMD5Password` which incorrectly claims that it accepts a plaintext password just like `IUsernameHashedPassword`.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_14"></a>|@jesstess commented|
|-|-|

(In [28225]) Branching to 'pb-md5-docstring-3056'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_15"></a>|@jesstess commented|
|-|-|

(In [28226]) checkMD5Password should say it works on a digest, not plaintext. Also,
the @return markup on a number of interfaces extending ICredentials
is hopefully now less confusing.

refs [#3056](https://github.com/twisted/twisted/issues/3056)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @jesstess|
|-|-|

Looks good.  Please merge.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_18"></a>|@jesstess set status to `closed`|
|-|-|

(In [28247]) Merge pb-md5-docstring-3056

Author: jesstess
Reviewer: exarkun
Fixes: [#3056](https://github.com/twisted/twisted/issues/3056)

IUsernameMD5Password.checkMD5Password now correctly say it works on a
digest, not plaintext. Also, the @return description on a number of
interfaces extending ICredentials has been clarified.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
